src/contexts/walletconnect.tsx

```diff
@@ -0,0 +1,200 @@
+import { BuildDeployContractTxData } from '../modals/SendModal/BuildDeployContractTx'
+import { BuildScriptTxData } from '../modals/SendModal/BuildScriptTx'
+import { BuildTransferTxData } from '../modals/SendModal/BuildTransferTx'
+import { extractErrorMsg } from '../utils/misc'
+import { useGlobalContext } from './global'
+
+export interface ContextType {
+  isWalletConnectModalOpen: boolean
+  setIsWalletConnectModalOpen: (isOpen: boolean) => void
+  walletConnect?: WalletConnectClient
+  setWalletConnect: Dispatch<SetStateAction<WalletConnectClient | undefined>>
+  dappTransactionData?:
+    | ['transfer', BuildTransferTxData]
+    | ['deploy-contract', BuildDeployContractTxData]
+    | ['script', BuildScriptTxData]
+    | undefined
+  requestEvent?: SessionTypes.RequestEvent
+  onError: (error: string) => void
+}
+
+export const initialContext: ContextType = {
+  isWalletConnectModalOpen: false,
+  setIsWalletConnectModalOpen: () => undefined,
+  walletConnect: undefined,
+  setWalletConnect: () => undefined,
+  dappTransactionData: undefined,
+  requestEvent: undefined,
+  onError: (error: string) => undefined
+}
+
+export const Context = createContext<ContextType>(initialContext)
+
+const respondError = (walletConnect: WalletConnectClient, requestEvent: SessionTypes.RequestEvent, error: string) => {
+  walletConnect.respond({
+    topic: requestEvent.topic,
+    response: {
+      id: requestEvent.request.id,
+      jsonrpc: '2.0',
+      error: {
+        code: -32000,
+        message: error
+      }
+    }
+  })
+}
+
+export const WalletConnectContextProvider: FC = ({ children }) => {
+  const { setTxModalType, settings } = useGlobalContext()
+  const { addresses } = useAddressesContext()
+  const [isWalletConnectModalOpen, setIsWalletConnectModalOpen] = useState(false)
+  const [walletConnect, setWalletConnect] = useState<WalletConnectClient>()
+  const [dappTransactionData, setDappTransactionData] = useState<ContextType['dappTransactionData']>()
+  const [requestEvent, setRequestEvent] = useState<SessionTypes.RequestEvent>()
+
+  const setTxData = useCallback(
+    (data: Exclude<ContextType['dappTransactionData'], undefined>) => {
+      setDappTransactionData(data)
+      setTxModalType(data[0])
+    },
+    [setDappTransactionData, setTxModalType]
+  )
+
+  const onError = useCallback(
+    (error: string): void => {
+      if (walletConnect && requestEvent) {
+        respondError(walletConnect, requestEvent, error)
+      }
+    },
+    [walletConnect, requestEvent]
+  )
+
+  useEffect(() => {
+    if (walletConnect === undefined) {
+      WalletConnectClient.init({
+        controller: true,
+
+        // TODO: add as an advanced settings option "WalletConnect Project Id"
+        projectId: '6e2562e43678dd68a9070a62b6d52207',
+        relayUrl: 'wss://relay.walletconnect.com',
+        metadata: {
+          name: 'Alephium Wallet',
+          description: 'Alephium Wallet',
```

Comment:
Should we clarify _"Alephium desktop wallet"_ now that we'll soon have a mobile as well (that could potentially have WalletConnect integration as well)?

Reply:
This is from @LeeAlephium 's code, I did not touch it. So I will leave it to you guys to change it

Reply:
No, @LeeAlephium closed it if you check the history

>@[lf94](https://github.com/lf94) lf94 closed this [20 days ago](https://github.com/alephium/desktop-wallet/pull/204#event-6559443468)

Reply:
We could target Lee's branch if it could help review

Reply:
Let's focus on answering the question rather than who needs to do what :slightly_smiling_face: 

> Should we clarify "Alephium desktop wallet" now that we'll soon have a mobile as well (that could potentially have WalletConnect integration as well)?

I think yes, since dApps may show this information on their side, and it could come from the desktop or mobile wallet.